File: js/lab/oo-split-context/facets.js
(function() {

let bottom = base._innards.bottom;
let address = base._innards.address;
let with_ctxt = base._innards.with_ctxt;

Set.prototype.union = function(elem) {
  let n = new Set(this);
  n.add(elem);
  return n;
};

function mk_facet(pc, v1, v2) {
  if (pc.size === 0)
    return v1;

  let [k, ...rest] = pc;
  rest = new Set(rest);

  if (k > 0)
    return facet(k, mk_facet(rest, v1, v2), v2);
  else
    return facet(k, v2, mk_facet(rest, v1, v2));
}

function facet(k, vh, vl) {
  return {
    k, vh, vl,
    eval_apply(C, v2) {
      if (C.pc.has(k)) {
        return vh.eval_apply(C, v2);
      }

      else if (pc.has(-k)) {
        return vl.eval_apply(C, v2);
      }

      else {
        let [C1, vh1] = vh.eval_apply(with_pc(C, C.pc.union(k)), v2);
        let [C2, vl1] = vl.eval_apply(with_pc(C1, C.pc.union(-k)), v2);
        return [ C2, mk_facet(k, vh1, vl1) ];
      }
    },

    eval_deref(C) {
      if (pc.has(k))
        return vh.eval_deref(C);
      else if (pc.has(-k))
        return vl.eval_deref(C);
      else
        return mk_facet(k, vh.eval_deref(C), vl.eval_deref(C));
    },

    eval_assign(C, v) {
      let C1 = vh.eval_assign(with_pc(C, C.pc.union(k)), v);
      let C2 = vl.eval_assign(with_pc(C1, C.pc.union(-k)), v);
      return C2;
    }
  };
}

function ref(e) {
  return {
    eval(C) {
      let [C1, v] = e.eval(C);
      let a = Object.keys(C1.σ).length;
      let C2 = with_ctxt(C1, {σ: Object.create(C1.σ)});
      C2.σ[a] = mk_facet(C1.pc, v, bottom);
      return [ C2, address(a) ];
    }
  };
}

function with_pc(C, pc) {
  return {__proto__: C, pc};
}

function interpretProgram(AST, env = {}, store = {}, pc = []) {
  return AST.eval({σ: env, θ: store, pc: new Set(pc)});
}

// Exports
this.facets = {
  __proto__: this.base,
  interpretProgram,
  ref
};

}());

File: java/aspectj/Facets.java
import java.util.*;

aspect Facets {
  static PC pc = null; //new PC().add(1);

  class PC {
    private List<Integer> vals;

    public PC() {
      vals = new ArrayList();
    }

    public boolean has(int v) {
      return vals.contains(v);
    }

    public PC add(int k) {
      PC p = new PC();
      p.vals = new ArrayList(this.vals);
      p.vals.add(k);
      return p;
    }

    public boolean isEmpty() { return vals.isEmpty(); }
    public int head() { return vals.get(0); }
    public PC tail() {
      PC p = new PC();
      p.vals = new ArrayList(vals.subList(1, vals.size()));
      return p;
    }
  }

  class Facet implements Value {
    int k;
    Value vh;
    Value vl;

    public Facet(int k, Value vh, Value vl) {
      this.k = k;
      this.vh = vh;
      this.vl = vl;
    }

    public Result evalApply(Store s, Value v) {
      if (pc.has(k))
        return vh.evalApply(s, v);
      else if (pc.has(-k))
        return vl.evalApply(s, v);
      else {
        PC old_pc = pc;
        pc = old_pc.add(k);
        Result rh = vh.evalApply(s, v);
        pc = old_pc.add(-k);
        Result rl = vl.evalApply(rh.store(), v);
        pc = old_pc;
        return new Result( rl.store(),
                           mkFacet(new PC().add(k),
                                   rh.value(),
                                   rl.value()));
      }
    }

    public Value evalDeref(Store s) {
      if (pc.has(k))
        return vh.evalDeref(s);
      else if (pc.has(-k))
        return vl.evalDeref(s);
      else
        return mkFacet(new PC().add(k), vh.evalDeref(s), vl.evalDeref(s));
    }

    public String toString() {
      return "<" + k + " ? " + vh + " : " + vl + ">";
    }
  }

  Value mkFacet(PC pc, Value v1, Value v2) {
    if (pc.isEmpty())
      return v1;

    int k = pc.head();

    if (k > 0)
      return new Facet(k, mkFacet(pc.tail(), v1, v2), v2);
    else
      return new Facet(k, v2, mkFacet(pc.tail(), v1, v2));
  }

  pointcut ref(Ref ref, Store s, Env env):
  call(Result eval(Store, Env))
    && target(ref)
    && args(s, env);

  Result around(Ref ref, Store s, Env env): ref(ref, s, env) {
    System.out.println("aspect");
    Result r1 = ref.e.eval(s, env);
    Address a = new Address(r1.store().length());
    Value v = mkFacet(pc, r1.value(), new Bottom());
    Store s2 = r1.store().add(v);
    return new Result(s2, a);
  }

  pointcut entry(): execution(void Test.main(String[]));

  before(): entry() {
    pc = Facets.aspectOf().new PC().add(1);
  }

}
